client/lib/wp/sync-handler/index.js

```diff
@@ -0,0 +1,210 @@
+ */
+export class SyncHandler {
+	/**
+	 * Create a SyncHandler instance
+	 *
+	 * @param {Object} [setup] - localforage sync setup
+	 * @param {Function} handler - wpcom handler
+	 *
+	 * @return {Function} sync-handler wrapper
+	 */
+	constructor( setup, handler ) {
+		if ( typeof setup === 'function' ) {
+			handler = setup;
+			setup = {};
+		}
+
+		this.setup = Object.assign( {}, defaults, setup );
+		this.reqHandler = handler;
+		return this.syncHandlerInstance( handler );
+	}
+
+	syncHandlerInstance( handler ) {
+		return ( params, callback ) => {
+			// detect and no-sync proxy connection request
+			if ( params.metaAPI && params.metaAPI.accessAllUsersBlogs ) {
+				debug( 'skip - non-sync -proxy-handler request detected' );
+				return this.reqHandler( params, callback );
+			}
+
+			const reqParams = Object.assign( {}, params );
+			const { path } = reqParams;
+
+			// whitelist barrier
+			if ( ! isWhitelisted( params ) ) {
+				debug( 'not whitelisted: skip %o request', path );
+				return handler( params, callback );
+			};
+
+			// response sent flag
+			let responseSent = false;
+
+			// generate an unique resource key
+			const key = this.generateKey( params );
+
+			debug( 'starting to get resource ...' );
+
+			/**
+			 * Send response immediately to the client
+			 * sendingt the given data
+			 *
+			 * @param {Object} localRecord - response stored locally
+			 */
+			const localResponse = localRecord => {
+				// let's be optimistic
+				if ( localRecord ) {
+					responseSent = true;
+
+					debug( '%o stored(%o). Let\'s be optimistic ...\n', path, localRecord );
+					//callback( null, localRecord.body );
+				} else {
+					debug( 'No data for %o\n', path );
+				}
+			};
+
+			/**
+			 * Handling Error getting data locally
+			 *
+			 * @param {Error} err - err trying to get the local record
+			 */
+			const recordErrorHandler = err => {
+				// @TODO improve error handling here
+				console.error( err );
+			};
+
+			/**
+			 * Fetch data from WP.com
+			 *
+			 * @return {Promise} promise
+			 */
+			const networkFetch = () => {
+				return new Promise( ( resolve, reject ) => {
+					handler( params, ( err, res ) => {
+						if ( err ) {
+							return reject( err );
+						}
+
+						resolve( res );
+					} );
+				} );
+			}
+
+			/**
+			 * Add/Override the data gotten from the
+			 * WP.com server-side response.
+			 * Run the double callback
+			 *
+			 * @param {Object} serverResponse - server response object
+			 */
+			const cacheResponse = serverResponse => {
+				// remove _headers from server response
+				delete serverResponse._headers;
+
+				let storingData = {
+					__sync: {
+						key,
+						synced: new Date().toString(),
+						syncing: false
+					},
+					body: serverResponse,
+					params: reqParams
+				};
+
+				this.storeRecord( key, storingData, err => {
+					if ( err ) {
+						console.error( err );
+
+						if ( ! responseSent ) {
+							callback( err, serverResponse );
```

Comment:
The thing is it's the second time that the callback is run. So if it's a storing error (localforage) I think is better just silence it. I think ins't useful run this second callback to the client.